### PR TITLE
fix(security): address CVE-2017-16138 

### PR DESCRIPTION
**Pull request by micoferdinand98**
Affected versions of mime are vulnerable to regular expression denial of service when a mime lookup is performed on untrusted user input.

Action Applied:
Updated to version 3.

mime
mime3 0

### CODE REVIEW

**Review by petermetz — CHANGES_REQUESTED** [review]
@micoferdinand98 You need to change the version of mime that is being used by existing dependencies within the monorepo packages. Right now you are just adding one additional declaration of it being a dependency instead.

**Comment by micoferdinand98** [comment]
> @micoferdinand98 You need to change the version of mime that is being used by existing dependencies within the monorepo packages. Right now you are just adding one additional declaration of it being a dependency instead.

image

Good Day Peter I would like to clarify on the Mime Version. is this enough for the dependencies on monorepo packages or should I downgrade the mime version?

**Comment by micoferdinand98** [comment]
> @micoferdinand98 You need to change the version of mime that is being used by existing dependencies within the monorepo packages. Right now you are just adding one additional declaration of it being a dependency instead.
M2

Good Day Peter I've inserted the updated mime version within the monorepo packages.

**Comment by micoferdinand98** [comment]
image
image

**Review by petermetz — CHANGES_REQUESTED** [review]
@micoferdinand98 
1. We don't need package-lock files (yarn.lock substitutes for it)
2. The generated files are being updated by the diff (deleted). Is that for a specific reason or just by accident?

**Review by petermetz — CHANGES_REQUESTED** [review]
@micoferdinand98 

> You need to change the version of mime that is being used by existing dependencies within the monorepo packages. Right now you are just adding one additional declaration of it being a dependency instead.

^^ This one still applies unfortunately.

```sh
$ npm ls mime
@hyperledger/cactus@ /home/peter/a/blockchain/cactus
├─┬ @hyperledger/cactus-cmd-api-server@1.0.0-rc.3 -> ./packages/cactus-cmd-api-server
│ ├─┬ express@4.17.1
│ │ ├─┬ send@0.17.1
│ │ │ └── mime@1.6.0
│ │ └─┬ serve-static@1.14.1
│ │   └─┬ send@0.17.1
│ │     └── mime@1.6.0
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-cmd-socket-server@1.0.0-rc.3 -> ./packages/cactus-cmd-socketio-server
│ ├─┬ express@4.16.4
│ │ └─┬ send@0.16.2
│ │   └── mime@1.4.1
│ └── mime@1.6.0
├─┬ @hyperledger/cactus-core@1.0.0-rc.3 -> ./packages/cactus-core
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-example-carbon-accounting-business-logic-plugin@1.0.0-rc.3 -> ./examples/cactus-example-carbon-accounting-business-logic-plugin
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-example-carbon-accounting-frontend@1.0.0-rc.3 -> ./examples/cactus-example-carbon-accounting-frontend
│ ├─┬ @angular-devkit/build-angular@12.2.0
│ │ ├─┬ less@4.1.1
│ │ │ └── mime@1.6.0
│ │ └─┬ webpack-dev-server@3.11.2
│ │   └─┬ webpack-dev-middleware@3.7.3
│ │     └── mime@2.6.0
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-example-supply-chain-backend@1.0.0-rc.3 -> ./examples/cactus-example-supply-chain-backend
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-example-supply-chain-business-logic-plugin@1.0.0-rc.3 -> ./examples/cactus-example-supply-chain-business-logic-plugin
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-consortium-manual@1.0.0-rc.3 -> ./packages/cactus-plugin-consortium-manual
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-htlc-eth-besu-erc20@1.0.0-rc.3 -> ./packages/cactus-plugin-htlc-eth-besu-erc20
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-htlc-eth-besu@1.0.0-rc.3 -> ./packages/cactus-plugin-htlc-eth-besu
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-keychain-azure-kv@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-azure-kv
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-keychain-google-sm@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-google-sm
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-keychain-memory-wasm@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-memory-wasm
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-keychain-memory@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-memory
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-ledger-connector-besu@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-besu
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-ledger-connector-fabric-socketio@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-fabric-socketio
│ ├─┬ express@4.15.5
│ │ └─┬ send@0.15.6
│ │   └── mime@1.3.4
│ └── mime@1.6.0
├─┬ @hyperledger/cactus-plugin-ledger-connector-fabric@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-fabric
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-ledger-connector-go-ethereum-socketio@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-go-ethereum-socketio
│ └── mime@1.6.0
├─┬ @hyperledger/cactus-plugin-ledger-connector-iroha@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-iroha
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-ledger-connector-quorum@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-quorum
│ └── mime@3.0.0 deduped
├─┬ @hyperledger/cactus-plugin-ledger-connector-sawtooth-socketio@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-sawtooth-socketio
│ └── mime@1.6.0
├─┬ karma@6.3.4
│ └── mime@2.6.0
└── mime@3.0.0
```

**Comment by micoferdinand98** [comment]
> @micoferdinand98
> 
> 1. We don't need package-lock files (yarn.lock substitutes for it)
> 2. The generated files are being updated by the diff (deleted). Is that for a specific reason or just by accident?

okay Peter I'll just get the package lock files removed.   The generated files was updated when I installed the mime in each packages.

**Comment by micoferdinand98** [comment]
> @micoferdinand98
> 
> > You need to change the version of mime that is being used by existing dependencies within the monorepo packages. Right now you are just adding one additional declaration of it being a dependency instead.
> 
> ^^ This one still applies unfortunately.
> 
> ```shell
> $ npm ls mime
> @hyperledger/cactus@ /home/peter/a/blockchain/cactus
> ├─┬ @hyperledger/cactus-cmd-api-server@1.0.0-rc.3 -> ./packages/cactus-cmd-api-server
> │ ├─┬ express@4.17.1
> │ │ ├─┬ send@0.17.1
> │ │ │ └── mime@1.6.0
> │ │ └─┬ serve-static@1.14.1
> │ │   └─┬ send@0.17.1
> │ │     └── mime@1.6.0
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-cmd-socket-server@1.0.0-rc.3 -> ./packages/cactus-cmd-socketio-server
> │ ├─┬ express@4.16.4
> │ │ └─┬ send@0.16.2
> │ │   └── mime@1.4.1
> │ └── mime@1.6.0
> ├─┬ @hyperledger/cactus-core@1.0.0-rc.3 -> ./packages/cactus-core
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-example-carbon-accounting-business-logic-plugin@1.0.0-rc.3 -> ./examples/cactus-example-carbon-accounting-business-logic-plugin
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-example-carbon-accounting-frontend@1.0.0-rc.3 -> ./examples/cactus-example-carbon-accounting-frontend
> │ ├─┬ @angular-devkit/build-angular@12.2.0
> │ │ ├─┬ less@4.1.1
> │ │ │ └── mime@1.6.0
> │ │ └─┬ webpack-dev-server@3.11.2
> │ │   └─┬ webpack-dev-middleware@3.7.3
> │ │     └── mime@2.6.0
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-example-supply-chain-backend@1.0.0-rc.3 -> ./examples/cactus-example-supply-chain-backend
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-example-supply-chain-business-logic-plugin@1.0.0-rc.3 -> ./examples/cactus-example-supply-chain-business-logic-plugin
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-consortium-manual@1.0.0-rc.3 -> ./packages/cactus-plugin-consortium-manual
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-htlc-eth-besu-erc20@1.0.0-rc.3 -> ./packages/cactus-plugin-htlc-eth-besu-erc20
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-htlc-eth-besu@1.0.0-rc.3 -> ./packages/cactus-plugin-htlc-eth-besu
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-keychain-azure-kv@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-azure-kv
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-keychain-google-sm@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-google-sm
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-keychain-memory-wasm@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-memory-wasm
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-keychain-memory@1.0.0-rc.3 -> ./packages/cactus-plugin-keychain-memory
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-besu@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-besu
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-fabric-socketio@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-fabric-socketio
> │ ├─┬ express@4.15.5
> │ │ └─┬ send@0.15.6
> │ │   └── mime@1.3.4
> │ └── mime@1.6.0
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-fabric@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-fabric
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-go-ethereum-socketio@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-go-ethereum-socketio
> │ └── mime@1.6.0
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-iroha@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-iroha
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-quorum@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-quorum
> │ └── mime@3.0.0 deduped
> ├─┬ @hyperledger/cactus-plugin-ledger-connector-sawtooth-socketio@1.0.0-rc.3 -> ./packages/cactus-plugin-ledger-connector-sawtooth-socketio
> │ └── mime@1.6.0
> ├─┬ karma@6.3.4
> │ └── mime@2.6.0
> └── mime@3.0.0
> ```

Okay got it now. instead of installing the mime package, I'll just manually replace the versions.

**Review by petermetz — APPROVED** [review]
@micoferdinand98 LGTM, thank you

**Comment by micoferdinand98** [comment]
> @micoferdinand98 LGTM, thank you

Thank you Peter

**Comment by takeutak** [comment]
@micoferdinand98 cc: @izuru0 
Thanks for your commit.  I'd like to confirm it works well before I approve your change (update the version of "express").  Would you wait for a few days?

**Review by izuru0 — CHANGES_REQUESTED** [review]
@micoferdinand98 thanks for updating.
but can you change ^4.17.3 to 4.17.3 i.e. specifying exact version?

**Comment by micoferdinand98** [comment]
> @micoferdinand98 thanks for updating. but can you change ^4.17.3 to 4.17.3 i.e. specifying exact version?

sure can do @izuru0 will change it now

**Comment by micoferdinand98** [comment]
> @micoferdinand98 cc: @izuru0 Thanks for your commit. I'd like to confirm it works well before I approve your change (update the version of "express"). Would you wait for a few days?

Yes it could wait, take your time @takeutak

**Review by izuru0 — APPROVED** [review]
LGTM

**Review by petermetz — CHANGES_REQUESTED** [review]
@micoferdinand98 Please fix the conflicts, other than that it's good to go from my side as well.

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json [hunk 1]
```diff
@@ -11,23 +11,19 @@
...
-    "debug": "4.1.1",
-    "express": "4.15.5",
-    "js-yaml": "3.14.1",
-    "jsonwebtoken": "8.5.1",
-    "log4js": "6.4.1",
-    "morgan": "1.10.0",
-    "serve-favicon": "2.4.5",
-    "shelljs": "0.8.5",
-    "socket.io": "4.4.1",
-    "xmlhttprequest": "1.8.0"
```
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:28]: @micoferdinand98 Was the change on this line intended or did the conflict resolution go sideways?

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json [hunk 2]
```diff
@@ -11,23 +11,19 @@
...
-  },
-  "devDependencies": {
```
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:30]: @micoferdinand98 Why was the `devDependencies` section removed?

File: packages/cactus-plugin-ledger-connector-fabric-socketio/package.json [hunk 3]
```diff
@@ -11,12 +11,12 @@
     "prepare-docker-build": "cp -af ../../yarn.lock ./dist/yarn.lock"
   },
   "dependencies": {
-    "@types/node": "14.17.32",
-    "body-parser": "1.17.2",
-    "config": "3.3.7",
-    "cookie-parser": "1.4.6",
-    "debug": "4.1.1",
-    "express": "4.15.5",
+    "@types/node": "^14.14.5",
```
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-fabric-socketio/package.json:14]: @micoferdinand98 Why go back to enabling auto-upgrade?

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-ledger-connector-fabric-socketio/package.json (hunk 3 above)
Comment (micoferdinand98) [reply packages/cactus-plugin-ledger-connector-fabric-socketio/package.json:14]: sorry peter it was changed when I was resolving the merge conflicts.

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json (hunk 2 above)
Comment (micoferdinand98) [reply packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:30]: It was the time rebasing the branch, I'll get it back now

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json (hunk 1 above)
Comment (micoferdinand98) [reply packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:28]: the resolution go sideways.

Comment (micoferdinand98) [reply packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:28]: the change happened when I squash other commits and resolved the merge conflict

File: packages/cactus-plugin-ledger-connector-fabric-socketio/package.json (hunk 3 above)
Comment (petermetz) [reply packages/cactus-plugin-ledger-connector-fabric-socketio/package.json:14]: @micoferdinand98 No worries. It still has auto-upgrade on it. The version of `@types/node` has also changed relative to `main`, I'm thinking that's also a conflict resolution mistake, but please confirm separately.

**Review by petermetz — CHANGES_REQUESTED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@micoferdinand98 The same issues still seem to be present in the diff as before.

**Review by petermetz — CHANGES_REQUESTED** [review]
@micoferdinand98 Cool, now the diff for the package.json files looks good. More nits below:
1. Fix the commit lint check - make sure to have a well formatted commit message and that it explains what was done and why
2. Fix the PR title check - I recommend having the PR title be the same thing as the first line of your commit message, helps with consistency
3. Do a `yarn install` to update the yarn.lock file, commit the changes that were made to the lock file (with --amend) so that it is integrated into the one commit of your PR. Then rebase onto the latest upstream/main and finally push force with lease.

**Review by petermetz — APPROVED** [review]
@micoferdinand98 LGTM, thank you!

**Pull request merged**